### PR TITLE
new librgw_admin_user

### PR DESCRIPTION
**Pull request by oritwas**
A library to create RGW user

### CODE REVIEW

File: src/include/rgw/librgw_admin_user.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+#ifndef LIB_RGW_ADMIN_USER_H
+#define LIB_RGW_ADMIN_USER_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define LIBRGW_ADMIN_USER_HVER_MAJOR 1
```
Comment (mattbenjamin) [inline src/include/rgw/librgw_admin_user.h]: HVER?

File: src/rgw/CMakeLists.txt [hunk 2]
```diff
@@ -270,6 +270,30 @@ set_target_properties(rgw PROPERTIES OUTPUT_NAME rgw VERSION 2.0.0
   SOVERSION 2)
 install(TARGETS rgw DESTINATION ${CMAKE_INSTALL_LIBDIR})
 
+set(librgw_admin_user_srcs
+  librgw_admin_user.cc
+  rgw_admin_user.cc)
+add_library(rgw_admin_user SHARED ${librgw_admin_user_srcs})
+target_link_libraries(rgw_admin_user PRIVATE
+  rgw_a
```
Comment (mattbenjamin) [inline src/rgw/CMakeLists.txt]: this brings in essentially everything in an instance, i.e., it's everything needed by librgw for file access;  at least the interface is small;  could this be reduced further, though?

**Comment by mattbenjamin** [comment]
Seems like a lot of symbols in the final link?

nm ./lib/librgw_admin_user.so | less
...
                 U close@@GLIBC_2.2.5
00000000007d98a0 b completed.6984
                 U curl_easy_cleanup
                 U curl_easy_getinfo
                 U curl_easy_init
...
00000000004b4150 T _ZN16rgw_cls_bi_entry11decode_jsonEP7JSONObjP15cls_rgw_obj_key
00000000004a34b0 W _ZN16rgw_cls_bi_entry6decodeERN4ceph6buffer4list8iteratorE
00000000004b3ca0 T _ZN16rgw_cls_bi_entry8get_infoEP15cls_rgw_obj_keyPhP25rgw_bucket_category_stats
00000000003a0e90 W _ZN16rgw_cls_bi_entryD1Ev
00000000003a0e90 W _ZN16rgw_cls_bi_entryD2Ev
00000000004aa340 W _ZN16rgw_cls_list_ret6decodeERN4ceph6buffer4list8iteratorE
00000000003165b0 T _ZN16rgw_datalog_info11decode_jsonEP7JSONObj
00000000003d5280 T _ZN16RGWDeleteObjTags17verify_permissionEv
00000000003ef2a0 W _ZN16RGWDeleteObjTags4nameB5cxx11Ev
00000000003d0ad0 T _ZN16RGWDeleteObjTags7executeEv
00000000003ee5e0 W _ZN16RGWDeleteObjTags7op_maskEv
00000000003ee5f0 W _ZN16RGWDeleteObjTags8get_typeEv
00000000003c0660 T _ZN16RGWDeleteObjTags8pre_execEv
00000000003ee540 W _ZN16RGWGetObj_Filter11fixup_rangeERlS0_
00000000002a2f70 W _ZN16RGWGetObj_Filter5flushEv
0000000000485d30 T _ZN16RGWGetRolePolicy10get_paramsEv
0000000000488b00 W _ZN16RGWGetRolePolicy4nameB5cxx11Ev
0000000000485ed0 T _ZN16RGWGetRolePolicy7executeEv
0000000000488a30 W _ZN16RGWGetRolePolicy8get_typeEv
...
Matt

**Comment by oritwas** [comment]
@mattbenjamin, the only library I could remove without getting undefined symbols is cls_replica_log_client.
I added librgw_common that contains the rgw source files needed for librgw_admin_user

**Comment by mattbenjamin** [comment]
@oritwas I don't think removing all the dependencies from rgw_a and librgw.so is what we want.  I'd like to be able to start multiple front-ends under librgw, for example.  It seems like librgw_common is a good idea in itself though--just rgw_a should I think include all the objects, whereas we don't need that in librgw_admin_user.  The librgw unit tests don't link due to missing symbols--that's trivial, it's just assuming some of what was taken out of rgw_a.

**Comment by oritwas** [comment]
@mattbenjamin , updated

**Comment by mattbenjamin** [comment]
@orit pasting a change that redefines rgw_common to be a cmake OBJECT library (re-used objects, no external static lib) (tested in my build env only)
```
commit 71da2dd0d6d665df6f2903d753cf7350710ac4bf
Author: Matt Benjamin <mbenjamin@redhat.com>
Date:   Mon Apr 23 14:23:16 2018 -0400

    object linkage
    
    Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

diff --git a/src/rgw/CMakeLists.txt b/src/rgw/CMakeLists.txt
index e2723f5ac4..fa1fbd872c 100644
--- a/src/rgw/CMakeLists.txt
+++ b/src/rgw/CMakeLists.txt
@@ -39,8 +39,6 @@ function(gperf_generate input output)
 endfunction()
 
 set(librgw_common_srcs
-  librgw_admin_user.cc
-  rgw_admin_user.cc
   rgw_acl.cc
   rgw_acl_s3.cc
   rgw_acl_swift.cc
@@ -67,6 +65,7 @@ set(librgw_common_srcs
   rgw_metadata.cc
   rgw_multi.cc
   rgw_multi_del.cc
+  rgw_otp.cc
   rgw_sync.cc
   rgw_data_sync.cc
   rgw_sync_module.cc
@@ -107,37 +106,19 @@ set(librgw_common_srcs
   rgw_crypt_sanitize.cc
   rgw_iam_policy.cc
 )
-add_library(rgw_common SHARED ${librgw_common_srcs})
+add_library(rgw_common OBJECT ${librgw_common_srcs})
 set(librgw_admin_user_srcs
   librgw_admin_user.cc
   rgw_admin_user.cc
 )
-target_link_libraries(rgw_common PRIVATE
-  librados
-  cls_rgw_client
-  cls_lock_client
-  cls_refcount_client
-  cls_log_client
-  cls_otp_client
-  cls_statelog_client
-  cls_timeindex_client
-  cls_version_client
-  cls_user_client
-  global
-  ${LIB_RESOLV}
-  ${CURL_LIBRARIES}
-  ${EXPAT_LIBRARIES})
-install(TARGETS rgw_common DESTINATION ${CMAKE_INSTALL_LIBDIR})
 
 set(rgw_a_srcs
-  ${librgw_common_srcs}
   rgw_auth_keystone.cc
   rgw_client_io.cc
   rgw_frontend.cc
   rgw_http_client_curl.cc
   rgw_loadgen.cc
   rgw_log.cchttps://start.fedoraproject.org/
-  rgw_otp.cc
   rgw_period_pusher.cc
   rgw_realm_reloader.cc
   rgw_realm_watcher.cc
@@ -169,7 +150,9 @@ if (WITH_RADOSGW_FCGI_FRONTEND)
   list(APPEND rgw_a_srcs rgw_fcgi.cc)
 endif()
 
-add_library(rgw_a STATIC ${rgw_a_srcs})
+add_library(rgw_a STATIC
+  ${rgw_a_srcs}
+  $<TARGET_OBJECTS:rgw_common>)
 
 add_dependencies(rgw_a civetweb_h)
 
@@ -302,12 +285,14 @@ set_target_properties(rgw PROPERTIES OUTPUT_NAME rgw VERSION 2.0.0
   SOVERSION 2)
 install(TARGETS rgw DESTINATION ${CMAKE_INSTALL_LIBDIR})
 
+add_library(rgw_admin_user SHARED
+  ${librgw_admin_user_srcs}
+  $<TARGET_OBJECTS:rgw_common>)
 
-add_library(rgw_admin_user SHARED ${librgw_admin_user_srcs})
 target_link_libraries(rgw_admin_user PRIVATE
-  rgw_common
   librados
   cls_rgw_client
+  cls_otp_client
   cls_lock_client
   cls_refcount_client
   cls_log_client
```

**Comment by oritwas** [comment]
jenkins test this please

**Comment by oritwas** [comment]
jenkins test this please

**Comment by mattbenjamin** [comment]
@oritwas needs a small rebase, but it's ready, I think
diff --git a/src/rgw/CMakeLists.txt b/src/rgw/CMakeLists.txt
index 9e50ee0477..adf712cef1 100644
--- a/src/rgw/CMakeLists.txt
+++ b/src/rgw/CMakeLists.txt
@@ -82,6 +82,7 @@ set(librgw_common_srcs
   rgw_cr_rest.cc
   rgw_object_expirer_core.cc
   rgw_op.cc
+  rgw_otp.cc
   rgw_policy_s3.cc
   rgw_quota.cc
   rgw_rados.cc
@@ -119,7 +120,6 @@ set(rgw_a_srcs
   rgw_realm_watcher.cc
   rgw_os_lib.cc
   rgw_process.cc
-  rgw_replica_log.cc
   rgw_request.cc
   rgw_rest_bucket.cc
   rgw_rest_config.cc
@@ -127,7 +127,6 @@ set(rgw_a_srcs
   rgw_rest_metadata.cc
   rgw_rest_opstate.cc
   rgw_rest_realm.cc
-  rgw_rest_replica_log.cc
   rgw_rest_swift.cc
   rgw_rest_usage.cc
   rgw_rest_user.cc
@@ -283,7 +282,7 @@ add_library(rgw_admin_user SHARED
 target_link_libraries(rgw_admin_user PRIVATE
   librados
   cls_rgw_client
-  cls_opt_client
+  cls_otp_client
   cls_lock_client
   cls_refcount_client
   cls_log_client

**Pull request merged**